internal/locate/region_cache.go

```diff
@@ -2169,88 +2251,81 @@ func (c *RegionCache) UpdateBucketsIfNeeded(regionID RegionVerID, latestBucketsV
 const cleanCacheInterval = time.Second
 const cleanRegionNumPerRound = 50
 
-// This function is expected to run in a background goroutine.
+// The returned function is expected to run in a background goroutine.
 // It keeps iterating over the whole region cache, searching for stale region
 // info. It runs at cleanCacheInterval and checks only cleanRegionNumPerRound
 // regions. In this way, the impact of this background goroutine should be
 // negligible.
-func (c *RegionCache) cacheGC() {
-	ticker := time.NewTicker(cleanCacheInterval)
-	defer func() {
-		c.wg.Done()
-		ticker.Stop()
-	}()
-
+func (c *RegionCache) gcRoundFunc(limit int) func(context.Context, time.Time) bool {
+	if limit < 1 {
+		limit = 1
+	}
 	beginning := newBtreeSearchItem([]byte(""))
-	iterItem := beginning
-	expired := make([]*btreeItem, cleanRegionNumPerRound)
-	remaining := make([]*Region, cleanRegionNumPerRound)
-	for {
-		select {
-		case <-c.ctx.Done():
-			return
-		case <-ticker.C:
-			count := 0
-			expired = expired[:0]
-			remaining = remaining[:0]
-
-			// Only RLock when checking TTL to avoid blocking other readers
-			c.mu.RLock()
-			ts := time.Now().Unix()
-			c.mu.sorted.b.AscendGreaterOrEqual(iterItem, func(item *btreeItem) bool {
-				if count > cleanRegionNumPerRound {
-					iterItem = item
-					return false
-				}
-				count++
-				if item.cachedRegion.isCacheTTLExpired(ts) {
-					expired = append(expired, item)
-				} else {
-					remaining = append(remaining, item.cachedRegion)
-				}
-				return true
-			})
-			c.mu.RUnlock()
-
-			// Reach the end of the region cache, start from the beginning
-			if count <= cleanRegionNumPerRound {
-				iterItem = beginning
+	cursor := beginning
+	expiredItems := make([]*btreeItem, limit)
+	needCheckRegions := make([]*Region, limit)
+
+	return func(_ context.Context, t time.Time) bool {
+		expiredItems = expiredItems[:0]
+		needCheckRegions = needCheckRegions[:0]
+		count, ts := 0, t.Unix()
+
+		// Only RLock when checking TTL to avoid blocking other readers
+		c.mu.RLock()
+		c.mu.sorted.b.AscendGreaterOrEqual(cursor, func(item *btreeItem) bool {
+			count++
+			if count > limit {
+				cursor = item
+				return false
 			}
+			if item.cachedRegion.isCacheTTLExpired(ts) {
+				expiredItems = append(expiredItems, item)
+			} else {
+				needCheckRegions = append(needCheckRegions, item.cachedRegion)
+			}
+			return true
+		})
+		c.mu.RUnlock()
 
-			// Clean expired regions
-			if len(expired) > 0 {
-				c.mu.Lock()
-				for _, item := range expired {
-					c.mu.sorted.b.Delete(item)
-					c.mu.removeVersionFromCache(item.cachedRegion.VerID(), item.cachedRegion.GetID())
-				}
-				c.mu.Unlock()
+		// Reach the end of the region cache, start from the beginning
+		if count < limit {
+			cursor = beginning
+		}
```

Comment:
If the count equals the limit, the start cursor for the next round is still the `beginning := newBtreeSearchItem([]byte(""))`, should we record the last cursor?

Reply:
Fixed